solr/core/src/test/org/apache/solr/cloud/SplitShardTest.java

```diff
@@ -174,6 +164,50 @@ public void testSplitFuzz() throws Exception {
     assertEquals("wrong range in s1_1", expected1, delta1);
   }
 
+  @Test
+  public void testWithChildDocuments() throws Exception {
+    CloudSolrClient solrClient = cluster.getSolrClient();
+    CollectionAdminRequest.createCollection(COLLECTION_NAME, "conf", 1, 1).process(solrClient);
+
+    cluster.waitForActiveCollection(COLLECTION_NAME, 1, 1);
+
+    List<SolrInputDocument> inputDocs = new ArrayList<>();
+    for (int idx = 0; idx < 10; ++idx) {
+      SolrInputDocument parent = new SolrInputDocument();
+      parent.addField("id", idx);
+      parent.addField("type_s", "parent");
+
+      // Child has different ID, so could be routed differently if used for routing.
+      SolrInputDocument child = new SolrInputDocument();
+      child.addField("id", idx + "_child");
+      child.addField("expected_parent_s", idx);
+
+      parent.addField("myChild", List.of(child));
+
+      inputDocs.add(parent);
+    }
+    solrClient.add(COLLECTION_NAME, inputDocs);
+
+    CollectionAdminRequest.SplitShard splitShard =
+        CollectionAdminRequest.splitShard(COLLECTION_NAME)
+            .setNumSubShards(2)
+            .setShardName("shard1");
+    splitShard.process(solrClient);
+    waitForState("Timed out waiting for shard split", COLLECTION_NAME, activeClusterShape(2, 2));
```

Comment:
Doesn't look like it does in this case. `class SplitShard extends AsyncCollectionAdminRequest`, and the logs show the wait having to try multiple times before predicate is satisfied
```
4141 INFO  (TEST-SplitShardTest.testWithChildDocuments-seed#[8A3CDF8B33E81E8F]) [n: c: s: r: x: t:] o.a.s.c.SolrCloudTestCase waitForState (splitshardtest-collection): Waiting for 2 active shards after split
4141 INFO  (TEST-SplitShardTest.testWithChildDocuments-seed#[8A3CDF8B33E81E8F]) [n: c: s: r: x: t:] o.a.s.c.SolrCloudTestCase active slice count: 1 expected: 2
4142 INFO  (watches-39-thread-1) [n: c: s: r: x: t:] o.a.s.c.SolrCloudTestCase active slice count: 1 expected: 2
4143 INFO  (TEST-SplitShardTest.testWithChildDocuments-seed#[8A3CDF8B33E81E8F]) [n: c: s: r: x: t:] o.a.s.c.SolrCloudTestCase active slice count: 1 expected: 2
4143 INFO  (watches-39-thread-1) [n: c: s: r: x: t:] o.a.s.c.SolrCloudTestCase active slice count: 1 expected: 2
4143 INFO  (TEST-SplitShardTest.testWithChildDocuments-seed#[8A3CDF8B33E81E8F]) [n: c: s: r: x: t:] o.a.s.c.SolrCloudTestCase active slice count: 1 expected: 2
4143 INFO  (TEST-SplitShardTest.testWithChildDocuments-seed#[8A3CDF8B33E81E8F]) [n: c: s: r: x: t:] o.a.s.c.SolrCloudTestCase active slice count: 1 expected: 2
4242 INFO  (zkCallback-40-thread-1) [n: c: s: r: x: t:] o.a.s.c.c.ZkStateReader A cluster state change: [WatchedEvent state:SyncConnected type:NodeDataChanged path:/collections/splitshardtest-collection/state.json zxid: 227] for collection [splitshardtest-collection] has occurred - updating... (live nodes size: [1])
4242 INFO  (zkCallback-24-thread-2) [n: c: s: r: x: t:] o.a.s.c.c.ZkStateReader A cluster state change: [WatchedEvent state:SyncConnected type:NodeDataChanged path:/collections/splitshardtest-collection/state.json zxid: 227] for collection [splitshardtest-collection] has occurred - updating... (live nodes size: [1])
4243 INFO  (watches-39-thread-1) [n: c: s: r: x: t:] o.a.s.c.SolrCloudTestCase active slice count: 2 expected: 2
4243 INFO  (watches-39-thread-1) [n: c: s: r: x: t:] o.a.s.c.SolrCloudTestCase active replica count: 2 expected replica count: 2
```

I can set `splitShard.setWaitForFinalState(true);`, but I think I still like having this verification step here since I've seen too many tests have false positives due to faulty premises, this helps future proof it.

Reply:
AsyncCollectionAdminRequest just means it's *capable* of asynchronous options if you choose to supply an async ID (you didn't).  Any way, no big deal.  waitForFinalState would probably work.  I just find all these waitFor....State calls in our tests excessive and sometimes needless.

Reply:
>  if you choose to supply an async ID (you didn't)

Good point, but it's evidently still not fully processing as the above logs show. And I just tried a more explicit test as follows, which  fails as per screenshot.

```java
    splitShard.process(solrClient);
    ZkStateReader zkStateReader = cluster.getZkStateReader();
    Collection<Slice> activeSlices = zkStateReader.getClusterState().getCollection(COLLECTION_NAME).getActiveSlices();
    assertEquals("Expected shard slices", activeSlices.size(), 2);
    ```

   
![image](https://github.com/user-attachments/assets/94e91218-0bf9-4d70-a6d0-bf456ef7b6c1)


~Do you know how `WAIT_FOR_FINAL_STATE` is used? I see that in the ShardSplitter it doesn't appear to be used directly, just passed on to it's own calls. So far I haven't followed those to learn more.~
EDIT: Actually upon multiple runs, it can still fail this assertion with  `setWaitForFinalState(true)`.

Reply:
Indeed it's possible that cluster.getZkStateReader() is not yet up to date but it doesn't matter since this is the client-side ZkStateReader (from a CloudSolrClient).  For this test, I don't think the client-side state view is relevant.  We're not testing CloudSolrClient in this test but the Solr-side behavior.

Reply:
> client-side ZkStateReader (from a CloudSolrClient)

derp 🤦 .

Well I was about to surrender, but instead put together one last sally :)

Inside the verification step, I added a log statement on the docs shard. `log.info("doc shard: "+doc.get("[shard]"));`
In the first ~3 runs this showed expected data  `shard1_0` etc, but eventually I git a run that showed the docs still on the original shard (attached pic). Though based on the logs above that, `Update shard state shard1_1 to active`, I'm guessing this is more of a visibility issue than a shard state issue and aforementioned wait wouldn't have prevented this. What could prevent this, a forced commit?

![image](https://github.com/user-attachments/assets/89b9c275-ba7c-47a5-8975-d35aaa97f972)

Reply:
Thanks for entertaining this line of inquiry and sharing your results but I don't want to press this matter at all.  I give up very easily on this matter; like 99% of tests do some waitForState like action after SolrCloud commands :-/

I suppose after a shard split, this wait step is necessary for a test if the test is ensuring certain separations post-split since otherwise, as you see, it's possible for the query to land on the parent shard still.  I thought such a query would get re-routed to the sub-shard but I guess not.  I have a clue where but ... don't worry about it.

Reply:
Okay sounds good. 

I was a little worried that, based on the last screenshot logs, this could still happen even with the wait, but it doesn't appear to, after 15 runs, so that's good.

```
for i in (seq 1 15)                                                                                                                                                                                         
  ./gradlew -p solr/core test --tests "*testWithChildDocuments*"
end
```

Reply:
Great! PS: You may enjoy to learn about the «beast ing» feature described in a sub chapter here https://github.com/apache/solr/blob/main/help/tests.txt